# Intro - Homework Assignment

## Question 1

What's the version of NumPy that you installed?

In [1]:
import numpy as np

np.__version__

'1.23.2'

### Getting the data

For this homework, we'll use the Car price dataset.
Download it from [here](https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv)

In [2]:
from pathlib import Path
import wget

url = 'https://raw.githubusercontent.com/alexeygrigorev/mlbookcamp-code/master/chapter-02-car-price/data.csv'
data_file = Path('data.csv')

if not data_file.exists():
    wget.download(url)

## Question 2

How many records are in the dataset?

Here you need to specify the number of rows.

In [3]:
import pandas as pd

df = pd.read_csv('data.csv')
df.shape

(11914, 16)

- 16
- 6572
- **11914**
- 18990

## Question 3

Who are the most popular car manufacturers (top-3) according to the dataset?

In [4]:
df.groupby('Make')['Popularity'].count().sort_values(ascending=False).head(3)

Make
Chevrolet     1123
Ford           881
Volkswagen     809
Name: Popularity, dtype: int64

- Chevrolet, Volkswagen, Toyota
- Chevrolet, Ford, Toyota
- Ford, Volkswagen, Toyota
- **Chevrolet, Ford, Volkswagen**

## Question 4

What's the number of unique Audi car models in the dataset?

In [5]:
df[df['Make'] == 'Audi']['Model'].unique()

array(['100', '200', '80', '90', 'A3', 'A4 allroad', 'A4', 'A5', 'A6',
       'A7', 'A8', 'allroad quattro', 'allroad', 'Cabriolet', 'Coupe',
       'Q3', 'Q5', 'Q7', 'R8', 'RS 4', 'RS 5', 'RS 6', 'RS 7', 'S3', 'S4',
       'S5', 'S6', 'S7', 'S8', 'SQ5', 'TT RS', 'TT', 'TTS', 'V8'],
      dtype=object)

In [6]:
len(df[df['Make'] == 'Audi']['Model'].unique())

34

- 3
- 16
- 26
- **34**

## Question 5

How many columns in the dataset have missing values?

In [6]:
df.isna().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

In [8]:
sum([1 for val in list(df.isna().sum().values) if val != 0])

5

- **5**
- 6
- 7
- 8

## Question 6

* Find the median value of "Engine Cylinders" column in the dataset.

In [9]:
df['Engine Cylinders'].median()

6.0

* Next, calculate the most frequent value of the same "Engine Cylinders".

In [10]:
df['Engine Cylinders'].mode()[0]

4.0

* Use the `fillna` method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step.

In [11]:
df['Engine Cylinders'].fillna(df['Engine Cylinders'].mode()[0], inplace=True)
df['Engine Cylinders'].isna().sum()

0

* Now, calculate the median value of "Engine Cylinders" once again.

Has it changed?

In [12]:
df['Engine Cylinders'].median()

6.0

> Hint: refer to existing `mode` and `median` functions to complete the task.

- Yes
- **No**

## Question 7

* Select all the "Lotus" cars from the dataset.

In [13]:
lotus_df = df[df['Make'] == 'Lotus']
lotus_df.head()

Make  Model  Year             Engine Fuel Type  Engine HP  \
3912  Lotus  Elise  2009  premium unleaded (required)      189.0   
3913  Lotus  Elise  2009  premium unleaded (required)      218.0   
3914  Lotus  Elise  2009  premium unleaded (required)      189.0   
3915  Lotus  Elise  2010  premium unleaded (required)      189.0   
3916  Lotus  Elise  2010  premium unleaded (required)      218.0   

      Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
3912               4.0            MANUAL  rear wheel drive              2.0   
3913               4.0            MANUAL  rear wheel drive              2.0   
3914               4.0            MANUAL  rear wheel drive              2.0   
3915               4.0            MANUAL  rear wheel drive              2.0   
3916               4.0            MANUAL  rear wheel drive              2.0   

              Market Category Vehicle Size Vehicle Style  highway MPG  \
3912  Exotic,High-Performance      Compact   Convertible           27   
3913  Exotic,High-Performance      Compact   Convertible           26   
3914  Exotic,High-Performance      Compact   Convertible           27   
3915  Exotic,High-Performance      Compact   Convertible           27   
3916  Exotic,High-Performance      Compact   Convertible           26   

      city mpg  Popularity   MSRP  
3912        21         613  43995  
3913        20         613  54990  
3914        21         613  47250  
3915        21         613  47250  
3916        20         613  54990

* Select only columns "Engine HP", "Engine Cylinders".

In [14]:
lotus_engines_df = lotus_df[['Engine HP', 'Engine Cylinders']]
lotus_engines_df.head()

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3914      189.0               4.0
3915      189.0               4.0
3916      218.0               4.0

* Now drop all duplicated rows using `drop_duplicates` method (you should get a dataframe with 9 rows).

In [19]:
lotus_engines_df_unique = lotus_engines_df.drop_duplicates()
lotus_engines_df_unique

Engine HP  Engine Cylinders
3912      189.0               4.0
3913      218.0               4.0
3918      217.0               4.0
4216      350.0               8.0
4257      400.0               6.0
4259      276.0               6.0
4262      345.0               6.0
4292      257.0               4.0
4293      240.0               4.0

In [20]:
lotus_engines_df_unique.shape

(9, 2)

* Get the underlying NumPy array. Let's call it `X`.

In [21]:
X = lotus_engines_df_unique.values
X

array([[189.,   4.],
       [218.,   4.],
       [217.,   4.],
       [350.,   8.],
       [400.,   6.],
       [276.,   6.],
       [345.,   6.],
       [257.,   4.],
       [240.,   4.]])

* Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.

In [33]:
np.matmul(X.T, X)

array([[7.31684e+05, 1.34100e+04],
       [1.34100e+04, 2.52000e+02]])

In [22]:
XTX = np.dot(X.T, X)
XTX

array([[7.31684e+05, 1.34100e+04],
       [1.34100e+04, 2.52000e+02]])

* Invert `XTX`.

In [23]:
XTXi = np.invert(XTX.astype(np.int64))

In [24]:
XTXi

array([[-731685,  -13411],
       [ -13411,    -253]])

* Create an array `y` with values `[1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800]`.

In [25]:
y = np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800])
y

array([1100,  800,  750,  850, 1300, 1000, 1000, 1300,  800])

* Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.

In [26]:
XTX.shape

(2, 2)

In [27]:
X.shape

(9, 2)

In [28]:
y.shape

(9,)

In [29]:
np.linalg.inv(XTX)

array([[ 5.53084235e-05, -2.94319825e-03],
       [-2.94319825e-03,  1.60588447e-01]])

In [31]:
w = np.dot(np.dot(np.linalg.inv(XTX), X.T), y)
w

array([  4.59494481, -63.56432501])

* What's the value of the first element of `w`?

In [32]:
w[0]

4.594944810094551

> **Note**: You just implemented linear regression. We'll talk about it in the next lesson.

- -0.0723
- **4.5949**
- 31.6537
- 63.5643